File: python/semantic_kernel/template_engine/blocks/val_block.py
# Copyright (c) Microsoft. All rights reserved.

import logging
from re import S, compile
<<<<<<< Updated upstream
<<<<<<< head
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
=======
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
<<<<<<< main
=======
<<<<<<< main
>>>>>>> origin/main
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
<<<<<<< main
>>>>>>> Stashed changes
=======
<<<<<<< main
>>>>>>> origin/main
from typing import TYPE_CHECKING, Any, ClassVar

from pydantic import model_validator

from semantic_kernel.exceptions import ValBlockSyntaxError
<<<<<<< Updated upstream
<<<<<<< head
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
from semantic_kernel.template_engine.blocks.block import Block
=======
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
<<<<<<< main
=======
=======
from typing import TYPE_CHECKING, Any, ClassVar, Optional, Tuple

from pydantic import model_validator

>>>>>>> ms/small_fixes
>>>>>>> origin/main
from semantic_kernel.template_engine.blocks.block import Block
from semantic_kernel.template_engine.blocks.block_errors import ValBlockSyntaxError
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
=======
from typing import TYPE_CHECKING, Any, ClassVar, Optional, Tuple

from pydantic import model_validator

>>>>>>> ms/small_fixes
from semantic_kernel.template_engine.blocks.block import Block
from semantic_kernel.template_engine.blocks.block_errors import ValBlockSyntaxError
>>>>>>> origin/main
=======
>>>>>>> Stashed changes
from semantic_kernel.template_engine.blocks.block_types import BlockTypes

if TYPE_CHECKING:
    from semantic_kernel.functions.kernel_arguments import KernelArguments
    from semantic_kernel.kernel import Kernel

logger: logging.Logger = logging.getLogger(__name__)

VAL_BLOCK_REGEX = r"^(?P<quote>[\"'])(?P<value>.*)(?P=quote)$"

VAL_BLOCK_MATCHER = compile(VAL_BLOCK_REGEX, flags=S)


class ValBlock(Block):
    """Create a value block.

    A value block is used to represent a value in a template.
    It can be used to represent any characters.
    It needs to start and end with the same quote character,
    can be both single or double quotes, as long as they are not mixed.

    Examples:
        'value'
        "value"
        'value with "quotes"'
        "value with 'quotes'"

    Args:
        content - str : The content of the value block.
        value - str: The value of the block.
        quote - str: The quote used to wrap the value.

    Raises:
        ValBlockSyntaxError: If the content does not match the value block syntax.

    """

    type: ClassVar[BlockTypes] = BlockTypes.VALUE
<<<<<<< Updated upstream
<<<<<<< head
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
    value: str | None = ""
    quote: str | None = "'"
=======
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
<<<<<<< main
    value: str | None = ""
    quote: str | None = "'"
=======
<<<<<<< Updated upstream
=======
>>>>>>> origin/main
=======
>>>>>>> Stashed changes
<<<<<<< main
    value: str | None = ""
    quote: str | None = "'"
=======
    value: Optional[str] = ""
    quote: Optional[str] = "'"
>>>>>>> ms/small_fixes
<<<<<<< Updated upstream
<<<<<<< head
>>>>>>> origin/main
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
=======
>>>>>>> Stashed changes
>>>>>>> origin/main

    @model_validator(mode="before")
    @classmethod
    def parse_content(cls, fields: Any) -> Any:
        """Parse the content and extract the value and quote.

        The parsing is based on a regex that returns the value and quote.
        if the 'value' is already present then the parsing is skipped.
        """
        if isinstance(fields, Block) or "value" in fields:
            return fields
        content = fields.get("content", "").strip()
        matches = VAL_BLOCK_MATCHER.match(content)
        if not matches:
            raise ValBlockSyntaxError(content=content)
        if value := matches.groupdict().get("value"):
            fields["value"] = value
        if quote := matches.groupdict().get("quote"):
            fields["quote"] = quote
        return fields

<<<<<<< Updated upstream
<<<<<<< head
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
    def render(self, *_: "Kernel | KernelArguments | None") -> str:
        """Render the value block."""
        return self.value or ""
=======
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
<<<<<<< main
    def render(self, *_: "Kernel | KernelArguments | None") -> str:
        """Render the value block."""
        return self.value or ""
=======
<<<<<<< Updated upstream
=======
>>>>>>> origin/main
=======
>>>>>>> Stashed changes
<<<<<<< main
    def render(self, *_: "Kernel | KernelArguments | None") -> str:
        """Render the value block."""
        return self.value or ""
=======
    def render(self, *_: Tuple["Kernel", Optional["KernelArguments"]]) -> str:
        return self.value
>>>>>>> ms/small_fixes
<<<<<<< Updated upstream
<<<<<<< head
>>>>>>> origin/main
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
<<<<<<< Updated upstream
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
>>>>>>> Stashed changes
=======
=======
>>>>>>> Stashed changes
>>>>>>> origin/main

File: nodejs/semantic-kernel/template-engine/blocks/val-block.ts
import { ValBlockSyntaxError } from '../../exceptions/template-engine-exceptions'
import { KernelArguments } from '../../functions/kernel-arguments'
import { Kernel } from '../../kernel'
import { Block, BlockConfig } from './block'
import { BlockTypes } from './block-types'

const VAL_BLOCK_REGEX = /^(?<quote>["'])(?<value>.*)(?=\k<quote>)["']$/s

/**
 * Configuration for a value block.
 */
export interface ValBlockConfig extends BlockConfig {
  /**
   * The value of the block.
   */
  value?: string | null

  /**
   * The quote used to wrap the value.
   */
  quote?: string | null
}

/**
 * Create a value block.
 *
 * A value block is used to represent a value in a template.
 * It can be used to represent any characters.
 * It needs to start and end with the same quote character,
 * can be both single or double quotes, as long as they are not mixed.
 *
 * Examples:
 *   'value'
 *   "value"
 *   'value with "quotes"'
 *   "value with 'quotes'"
 */
export class ValBlock extends Block {
  /**
   * The type of the block.
   */
  public static override readonly type: BlockTypes = BlockTypes.VALUE

  /**
   * The value of the block.
   */
  public value: string | null

  /**
   * The quote used to wrap the value.
   */
  public quote: string | null

  /**
   * Creates a new ValBlock instance.
   * @param config - The configuration for the block.
   * @throws ValBlockSyntaxError if the content does not match the value block syntax.
   */
  constructor(config: ValBlockConfig) {
    super(config)

    // If value is already provided, use it directly
    if (config.value !== undefined) {
      this.value = config.value
      this.quote = config.quote ?? "'"
    } else {
      // Parse the content to extract value and quote
      const parsed = this.parseContent(this.content)
      this.value = parsed.value
      this.quote = parsed.quote
    }
  }

  /**
   * Parse the content and extract the value and quote.
   * @param content - The content to parse.
   * @returns An object containing the value and quote.
   * @throws ValBlockSyntaxError if the content does not match the expected format.
   */
  private parseContent(content: string): { value: string | null; quote: string | null } {
    const matches = VAL_BLOCK_REGEX.exec(content)
    if (!matches || !matches.groups) {
      throw new ValBlockSyntaxError(content)
    }

    const value = matches.groups['value'] || null
    const quote = matches.groups['quote'] || null

    return { value, quote }
  }

  /**
   * Render the value block.
   * @param _kernel - The kernel instance (unused).
   * @param _arguments - The kernel arguments (unused).
   * @returns The value of the block.
   */
  public render(_kernel?: Kernel, _arguments?: KernelArguments | null): string {
    return this.value || ''
  }
}
